datasync/chaindatafetcher/kafka/consumer.go

```diff
@@ -101,20 +158,131 @@ func (c *Consumer) Cleanup(s sarama.ConsumerGroupSession) error {
 	return nil
 }
 
+// insertSegment inserts the given segment to the given buffer.
+// Assumption:
+//  1. it is guaranteed that the order of segments is correct.
+//  2. the inserted messages may be duplicated.
+//
+// We can consider the following cases.
+// case1. new segment with index 0 is inserted into newly created segment slice.
+// case2. new consecutive segment is inserted into the right position.
+// case3. duplicated segment is ignored.
+// case4. new sparse segment shouldn't be given, so return an error.
+func insertSegment(segment *Segment, buffer [][]*Segment) ([][]*Segment, error) {
+	for idx, segments := range buffer {
+		length := len(segments)
+		if length > 0 && segments[0].key == segment.key {
+			// the segment is already inserted to buffer.
+			if segment.index < uint64(length) {
+				logger.Warn("the message is duplicated", "segment", segment)
+				return buffer, nil
+			}
+
+			// the segment is inserted.
+			if segment.index == uint64(length) {
+				buffer[idx] = append(segments, segment)
+				return buffer, nil
+			}
+
+			// there is a missing segment which should not be occurred.
+			return buffer, errors.New(wrongSegmentErrorMsg)
+		}
+	}
+
+	if segment.index == 0 {
+		// create a segment slice and append it.
+		buffer = append(buffer, []*Segment{segment})
+	} else {
+		// the segment may be already handled.
+		logger.Warn("the message may be inserted already. drop the segment", "segment", segment)
+	}
+	return buffer, nil
+}
+
+// handleBufferedMessages handles all consecutive complete messages in the buffer.
+func (c *Consumer) handleBufferedMessages(buffer [][]*Segment) ([][]*Segment, error) {
+	// base case
+	if len(buffer) <= 0 {
+		return buffer, nil
+	}
+
+	// if any message exists in the buffer
+	oldestMsg, firstSegment, total := buffer[0], buffer[0][0], len(buffer[0])
+	if total > 0 {
+		if uint64(total) != firstSegment.total {
+			// not ready for assembling messages
+			return buffer, nil
+		}
+
+		// ready for assembling message
+		var msgBuffer []byte
+		for _, segment := range oldestMsg {
+			msgBuffer = append(msgBuffer, segment.value...)
+		}
+		msg := &sarama.ConsumerMessage{
+			Key:   []byte(firstSegment.key),
+			Value: msgBuffer,
+		}
+
+		f, ok := c.handlers[firstSegment.orig.Topic]
+		if !ok {
+			return buffer, fmt.Errorf("%v: %v", noHandlerErrorMsg, msg.Topic)
+		}
+
+		if err := f(msg); err != nil {
+			return buffer, err
+		}
+
+		buffer = buffer[1:]
+	}
+	return c.handleBufferedMessages(buffer)
```

Comment:
How about changing this recursive structure to a repeat structure?

Reply:
Reflected. d917100